[4.0] Making CategoryNode and MenuItem ImmutableNodes

Hackwar opened the pull request:
In #23364 we added a node interface for object nodes and in this PR I'm making CategoryNode and MenuItem implement those interfaces. That should then make it possible to drop `\Joomla\CMS\Menu\Node` and we can maybe replace some nasty stuff where we build menus or similar stuff with convoluted code with proper tree traversal instead.

I added a trait for such node systems and changed CategoryNode and MenuItem to use those and a few fixes and improvements. I then also changed the menu item discovery of routing. This should improve the parseing quite a bit, since we don't have to run through all menu items.

I also fixed a few docblocks...

Looking forward to your feedback. Next thing would be to change mod_menu to use such a tree iterator to render the menu instead of what we have right now.

## Testinstructions
The idea is, that this does not change anything visibly for the user. Take your testsite, preferably with some deeply nested menu items and lots of menu items. Check that the links all work and maybe how long it takes for Joomla to render a page. Apply this patch and see that the URLs stay the same and still all work and hopefully also see a slight performance improvement for page rendering by Joomla. The performance improvement should only be noticeable for really large sites with several thousand menu items.

Hackwar committed:
Making CategoryNode and MenuItem ImmutableNodes

Hackwar committed:
Adding ImmutableNodeTrait, NodeTrait
Optimising CategoryNode, MenuItem
Improving menu item discovery in routing

brianteeman commented on libraries/src/Tree/NodeTrait.php:
```diff
@@ -0,0 +1,108 @@
+		{
+			end($this->_parent->_children);
+			$this->_leftSibling = prev($this->_parent->_children);
+			$this->_leftSibling->_rightsibling = & $this;
```
Don't know if it matters but the case if rightsibling is different

leftSibling
rightsibling

Hackwar replied:
Doesn't make a difference, but I'll fix it.

Hackwar committed:
Fixing typo

Hackwar commented:
FYI: If we had running unittests, they would fail because of backwards incompatible changes here. They aren't major changes, but still the behavior is slightly different.

Hackwar committed:
Fixing reference bug

Hackwar committed:
Merge branch 'j4nodes' of github.com:Hackwar/joomla-cms into j4nodes

# Conflicts:
#	libraries/src/Tree/NodeTrait.php

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4nodes

Hackwar committed:
Rewriting backend menu to use MenuItem
Fixing bug in NodeInterface/NodeTrait

Hackwar committed:
Codestyle

Hackwar committed:
Removing unused use statement

Hackwar commented:
I changed the backend menu to use the MenuItem class and re-ordered a lot of code as well. This makes about a 1000 lines of code obsolete...

These changes had 3 goals:
1. Make the Joomla\CMS\Menu\Node class obsolete, since it is a copy of MenuItem of this PR.
2. Reduce the code necessary for the backend menu.
3. Make the code easier to understand.

I succeeded with 1. and as you can see reduced the code quite a bit, but it still isn't easy to understand. **sigh**

The current implementation works in a way that is very strange to me... I don't even really know how to explain how it works. I felt uncomfortable working with first flat lists, then converting those into a pseudo-tree, then converting that again into a "real" tree, but being unable to access that without going through the Tree class first... The new implementation now works with nodes of type MenuItem right from the start and then modifies and iterates the whole tree afterwards. The code still is pretty complex, but I hope that I reduced the complexity by a degree. Maybe from a 10 to a 9?

I also moved code from the Joomla\CMS\Menu\MenuHelper to the MenusHelper of com_menus, because we are already using code from that class in that whole system and I couldn't see a reason for having some code in the component and other code in the framework. Since this is highly application specific code, I decided to move this entirely to the components MenusHelper and to drop the framework class. This reduces the rendering of the backend menu from 10 files down to 3.

One detail that I feel a bit uncomfortable with is removing the Joomla\CMS\Menu\Node\* classes. The backend menu has special classes for each node type which contain some special handling and which I would consider a rather better implementation than what is in this PR. But I decided against those node-type-classes since it brings the backend menu closer to a copy of the frontend menu. At the same time, there is so much special logic in the code to render the menu, that a little bit more to acommodate for this wont change the overall impression of this noticeably.

The part of creating your own backend menu has to be tested again, but shouldn't be far off from what we expect.

@wilsonge I hope that raises the chance to merge this. If you want to know a different reason to make MenuItem implement the NodeInterface: With this change, you could dynamically extend the menu system during runtime. You could generate menu items from categories and insert them wherever you want by doing addChild(), etc. and it would automatically propagate through routing and menu displays without storing this in the DB. It would make the menu more flexible. ;-)

wilsonge committed:
Merge branch '4.0-dev' into j4nodes

wilsonge committed:
Merge branch '4.0-dev' into j4nodes

wilsonge commented on administrator/components/com_menus/Helper/MenusHelper.php:
```diff
@@ -367,7 +402,7 @@ public static function getMenuItems($menutype, $enabledOnly = false, $exclude =
 	 * @throws  \Exception
 	 *
-	 * @since   3.8.0
+	 * @since   __DEPLOY_VERSION__
```
Are this version actually changing from 3.8?

Hackwar replied:
You are right, this did not really change.

Hackwar replied:
Fixed it.

wilsonge commented on administrator/modules/mod_menu/Menu/CssMenu.php:
```diff
@@ -253,29 +245,31 @@ protected function check($items, Registry $params)
 		// Call preprocess for the menu items on plugins.
 		// Plugins should normally process the current level only unless their logic needs deep levels too.
-		$this->application->triggerEvent('onPreprocessMenuItems', array('com_menus.administrator.module', &$items, $this->params, $this->enabled));
+		$this->application->triggerEvent('onPreprocessMenuItems', array('com_menus.administrator.module', $children, $this->params, $this->enabled));
```
Is this going to cause a b/c break on existing plugins triggering on this? Just for the purposes of documenting it if it is

Hackwar replied:
yes and no. It principally is the same event with the same data, but the actual menu item objects are now different to the current ones. It depends on the plugin if it needs updates, since there is no `submenu` attribute anymore, but the methods to traverse the tree.

wilsonge replied:
OK. But it will need a few lines of docs to be safe. Honestly I have no clue if this is used but always better safe than sorry

Hackwar replied:
I hope that comment explains this enough?

Hackwar committed:
Reverting docblock change

Hackwar committed:
Revert another docblock change

Hackwar committed:
Fixing docblocks

wilsonge committed:
Merge branch '4.0-dev' into j4nodes

Hackwar commented:
Deprecation notices are added in #23842

Hackwar committed:
Improving comment for onPreprocessMenuItems

Hackwar committed:
Fixing issues with custom backend menus

wilsonge committed:
Merge branch '4.0-dev' into j4nodes

Hackwar committed:
Fixing notice

chmst commented:
I have tested this item :white_check_mark: successfully on cf51ece371304b7a7993b3c61f5c0a7ba296c513<br /><br />I have tested this successfully - from the user's point of view - without code inspection. 
I used 5 levels of categories. Some with articles and some without. 
In Backend and Frontend, the tested structure looks equal after patch and before patch and links work as expected.

The amount of categories was not sufficient to say anything about performance. 
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23452">issues.joomla.org/tracker/joomla-cms/23452</a>.</sub>

Hackwar committed:
Fixing typo

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4nodes

wilsonge committed:
Merge branch '4.0-dev' into j4nodes

wilsonge commented:
Thanks for being patient with this!
